components/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,357 @@
+                return false;
+            }
+        }
+        // Warming-up
+        return true;
+    }
+
+    // Not warming-up
+    return false;
+}
+
+/*!
+ * \brief Check minimum interval between CO2 reads
+ * \details
+ *      Not described in the datasheet, but it is the same frequency as the built-in LED blink.
+ * \retval true
+ *      Ready to call readCO2().
+ * \retval false
+ *      Conversion not completed.
+ */
+bool mhz19b_is_ready(mhz19b_dev_t *dev)
+{
+    // Minimum CO2 read interval (Built-in LED flashes)
+    if ((esp_timer_get_time() - dev->last_ts) > MHZ19B_READ_INTERVAL_MS) {
+        return true;
+    }
+
+    return false;
+}
+
+/*!
+ * \brief Read CO2 from sensor
+ * \retval <0
+ *      MH-Z19B response error codes.
+ * \retval 0..399 ppm
+ *      Incorrect values. Minimum value starts at 400ppm outdoor fresh air.
+ * \retval 400..1000 ppm
+ *      Concentrations typical of occupied indoor spaces with good air exchange.
+ * \retval 1000..2000 ppm
+ *      Complaints of drowsiness and poor air quality. Ventilation is required.
+ * \retval 2000..5000 ppm
+ *      Headaches, sleepiness and stagnant, stale, stuffy air. Poor concentration, loss of
+ *      attention, increased heart rate and slight nausea may also be present.\n
+ *      Higher values are extremely dangerous and cannot be measured.
+ */
+esp_err_t mhz19b_read_CO2(mhz19b_dev_t *dev, int16_t *co2)
+{
+    esp_err_t result;
+    // Send command "Read CO2 concentration"
+    result = mhz19b_send_command(dev, MHZ19B_CMD_READ_CO2, 0, 0, 0, 0, 0);
+
+    // Check result
+    if (result == ESP_OK) {
+        // 16-bit CO2 value in response Bytes 2 and 3
+        *co2 = (dev->buf[2] << 8) | dev->buf[3];
+        dev->last_ts = esp_timer_get_time();
+        dev->last_value = *co2;
+    }
+    return result;
+}
+
+/*!
+ * \brief Get firmware version (NOT DOCUMENTED)
+ * \details
+ *      This is an undocumented command, but most sensors returns ASCII "0430 or "0443".
+ * \param version
+ *      Returned character pointer to version (must be at least 5 Bytes)\n
+ *      Only valid when return is set to MHZ19B_RESULT_OK.
+ * \param versionLen
+ *      Number of characters including NULL of version buffer.
+ * \return
+ *      MH-Z19B response error codes.
+ */
+esp_err_t mhz19b_get_version(mhz19b_dev_t *dev, char *version, uint8_t version_len)
+{
+    esp_err_t result;
+
+    // Argument check
+    if (version_len < 5) {
+        return MHZ19B_RESULT_ARGUMENT_ERROR;
+    }
+
+    // Clear version
+    memset(version, 0, 5);
+
+    // Send command "Read firmware version" (NOT DOCUMENTED)
+    result = mhz19b_send_command(dev, MHZ19B_CMD_GET_VERSION, 0, 0, 0, 0, 0);
+
+    // Check result
+    if (result == ESP_OK) {
+        // Copy 4 ASCII characters to version array like "0443"
+        for (uint8_t i = 0; i < 4; i++) {
+            // Version in response Bytes 2..5
+            version[i] = dev->buf[i + 2];
+        }
+    }
+    return result;
+}
+
+/*!
```

Comment:
Please move all doxygen comments to the header. Source files (* .c) are not processed by the documentation generator.